rust/datafusion/src/physical_plan/csv.rs

```diff
@@ -106,13 +107,71 @@ impl<'a> CsvReadOptions<'a> {
     }
 }
 
+///  SourceReader represents where the data comes from.
+enum SourceReader {
+    /// The data comes from partitioned files
+    PartitionedFiles {
+        /// Path to directory containing partitioned files with the same schema
+        path: String,
+        /// The individual files under path
+        filenames: Vec<String>,
+    },
+
+    /// The data comes from anything impl Read trait
+    Reader(Mutex<Option<Box<dyn Read + Send + Sync + 'static>>>),
+}
+
+impl std::fmt::Debug for SourceReader {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SourceReader::PartitionedFiles { path, filenames } => f
+                .debug_struct("PartitionedFiles")
+                .field("path", path)
+                .field("filenames", filenames)
+                .finish()?,
+            SourceReader::Reader(_) => f.write_str("Reader")?,
+        };
+        Ok(())
+    }
+}
+
+impl Clone for SourceReader {
+    fn clone(&self) -> Self {
+        match self {
+            SourceReader::PartitionedFiles { path, filenames } => {
+                Self::PartitionedFiles {
+                    path: path.clone(),
+                    filenames: filenames.clone(),
+                }
+            }
+            SourceReader::Reader(_) => Self::Reader(Mutex::new(None)),
```

Comment:
This might cause some non trivial confusion -- namey that `Clone'`ing a SourceReader will *not* clone the underlying reader. Thus any `Clone`'d `CsvExec` won't be usable at at all (it will generate an error)

 I wonder if `CsvExec` really needs to be `Clone` at all -- like can we just remove the `Clone` derivation:

```
#[derive(Debug)]
pub struct CsvExec {
```

Reply:
I agree it's unnecessary for `CsvExec` to be `Clone`. But if we remove the `Clone` derivation, will it introduce a breaking change ? 

If the `CsvExec` is built from source files ( not from a reader ) , the `Clone` will act as expected.

Reply:
Removing `Clone` would be a breaking change I agree -- though I am not sure how many people `Clone` physical plans

Reply:
But `CsvExec::with_new_children` requires itself `Clone` ...